content/blog/2020-08-19-train-a-sequence-model-with-poutyne/index.en.md

```diff
@@ -0,0 +1,613 @@
+    def forward(self, padded_sequences_vectors, lengths):
+        """
+            Defines the computation performed at every call.
+
+            Shapes:
+                padded_sequences_vectors: batch_size * longest_sequence_length (padding), 300
+                lengths: batch_size
+
+        """
+        total_length = padded_sequences_vectors.shape[1]
+        pack_padded_sequences_vectors = pack_padded_sequence(padded_sequences_vectors, lengths.cpu(), batch_first=True)
+
+        lstm_out, self.hidden_state = self.lstm_network(pack_padded_sequences_vectors)
+        lstm_out, _ = pad_packed_sequence(lstm_out, batch_first=True, total_length=total_length)
+
+        tag_space = self.fully_connected_network(lstm_out) # shape: batch_size * longest_sequence_length, 8 (tag space)
+        return tag_space.transpose(-1, 1) # we need to transpose since it's a sequence # shape: batch_size * 8, longest_sequence_length
+
+full_network = FullNetWork(lstm_network, fully_connected_network)
+```
+
+## Summary
+
+We have created an LSTM network (`lstm_network`) and a fully connected network (`fully_connected_network`), and we use both
+components in the full network. The full network makes use of padded-packed sequences, 
+so we created the `pad_collate_fn` function to process the necessary work within the `DataLoader`. Finally, 
+we will load the data using the vectorizer (within the `DataLoader` using the `pad_collate` function). This means that the addresses will be represented by word embeddings. 
+Also, the address components will be converted into categorical value (from 0 to 7).
+
+
+## The Training
+
+Now that we have all the components for the network, let's define our optimizer (Stochastic Gradient Descent) 
+([SGD](https://en.wikipedia.org/wiki/Stochastic_gradient_descent)).
+
+```python
+optimizer = optim.SGD(full_network.parameters(), lr)
+```
+
+### Poutyne Experiment
+> Disclaimer: David is a dev on Poutyne, so we will present code using this framework. See the project [here](https://poutyne.org/).
+
+Let's create our experiment using Poutyne for automatically logging in the project root directory (`./`). We will also set
+the loss function and a batch metric (accuracy) to monitor the training.
+
+```python
+exp = Experiment("./", full_network, device=device, optimizer=optimizer,
+                 loss_function=cross_entropy, batch_metrics=["acc"])
+```
+
+Using our experiment, we can now launch the training as simply as
+
+```python
+exp.train(train_loader, valid_generator=valid_loader, epochs=epoch_number)
+```
+It will take around 6 minutes per epochs, so around an hour for the complete training.
+
+### Results
+The next figure shows the loss and the accuracy during our training (blue) and during our validation (orange) steps.
+After 10 epochs, we obtain a validation loss and accuracy of `0.01981` and `99.54701` respectively, which is pretty
+good for a first model. Also, we can see that our model did not seem to have overfitted.
```

Comment:
```suggestion
good for a first model. Also, since our training accuracy and loss closely match their respective validation values, our model does not appear to be overfitted on the training set.
```